core/src/main/scala/zio/jdbc/ZStatefulConnection.scala

```diff
@@ -0,0 +1,100 @@
+/*
+ * Copyright 2022 John A. De Goes and the ZIO Contributors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package zio.jdbc
+
+import zio._
+import zio.jdbc.ZStatefulConnection._
+
+import java.sql.Connection
+
+final class ZStatefulConnection(
+  private[jdbc] val underlying: Connection,
+  private[jdbc] val dirtyBits: Ref[Int],
+  private[jdbc] val state: Ref[State],
+  private[jdbc] val defaultIsolationLevel: TransactionIsolationLevel
+) {
```

Comment:
This actually needs to extend `Connection`, and you need to keep track of many dirty bits: anything which is "touched".

It's not enough to wrap connection and use this interface, because `access` in ZConnection will expose the raw underlying java Connection, which is used by a lot of code. So you have to do the dirty bits calculation at the level of the Java Connection.